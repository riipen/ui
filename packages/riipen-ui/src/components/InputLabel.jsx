import clsx from "clsx";
import PropTypes from "prop-types";
import React from "react";

import ThemeContext from "../styles/ThemeContext";
import withClasses from "../utils/withClasses";

import InputHint from "./InputHint";
import Typography from "./Typography";

class InputLabel extends React.Component {
  static propTypes = {
    /**
     * The content of the component.
     */
    children: PropTypes.node,

    /**
     * An array of custom CSS classes to apply.
     */
    classes: PropTypes.array,

    /**
     * Hint text to display under the label.
     */
    hint: PropTypes.node,

    /**
     * If true, an asterisk will be appended to the end of the label.
     */
    required: PropTypes.bool
  };

  static defaultProps = {
    classes: [],
    required: false
  };

  static contextType = ThemeContext;

  render() {
    const { children, classes, hint, required, ...other } = this.props;

    const theme = this.context;

    const className = clsx(classes);

    const marginBottom = hint ? 1 : 3;

    return (
      <React.Fragment>
        <label className={className} {...other}>
          <Typography color="inherit" variant="inherit">
            {children}
            {required && " *"}
          </Typography>
        </label>
        {hint && <InputHint>{hint}</InputHint>}
        <style jsx>{`
          label {
            color: ${theme.palette.text.secondary};
            display: inline-block;
            font-family: ${theme.typography.body1.fontFamily};
            font-size: 16px;
            font-weight: ${theme.typography.body1.fontWeight};
            letter-spacing: ${theme.typography.body1.letterSpacing};
            line-height: 1.4;
            margin-bottom: ${theme.spacing(marginBottom)}px;
          }
        `}</style>
      </React.Fragment>
    );
  }
}

export default withClasses(InputLabel);
