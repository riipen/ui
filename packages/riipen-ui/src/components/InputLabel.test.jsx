import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import React from "react";

import InputLabel from "./InputLabel.jsx";

describe("<InputLabel>", () => {
  it("renders without errors", () => {
    let error;
    try {
      mount(<InputLabel />);
    } catch (e) {
      error = e;
    }
    expect(error).toEqual(undefined);
  });

  it("renders correct snapshot", () => {
    const wrapper = mount(<InputLabel />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  describe("default props", () => {
    it("sets correct default props", () => {
      const defaultProps = new InputLabel().type.defaultProps;

      const wrapper = mount(<InputLabel />);

      const component = wrapper.find("InputLabel");
      expect(component.props().color).toEqual(defaultProps.color);
      expect(component.props().required).toEqual(defaultProps.required);
      expect(component.props().weight).toEqual(defaultProps.weight);
    });
  });

  describe("children prop", () => {
    it("renders given children inside the label and Typography element", () => {
      const child = <div>it's a label</div>;

      const wrapper = mount(<InputLabel children={child} />);

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .find("Typography")
          .contains(child)
      ).toEqual(true);
    });

    it("does not render label and Typography elements when children is not given", () => {
      const wrapper = mount(<InputLabel />);

      expect(wrapper.find("InputLabel").find("label")).toHaveLength(0);
      expect(wrapper.find("InputLabel").find("Typography")).toHaveLength(0);
    });
  });

  describe("classes prop", () => {
    it("applies all classes to the label element when children is given", () => {
      const classes = ["classOne"];
      const child = <div>it's a label</div>;

      const wrapper = mount(<InputLabel classes={classes} children={child} />);

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .hasClass(classes[0])
      ).toEqual(true);
    });

    it("appends higher order values to classes prop with withClass decorator", () => {
      const classes = ["riipen", "riipen-inputlabel"];

      const wrapper = mount(<InputLabel />);

      expect(
        wrapper
          .find("InputLabel")
          .props()
          .classes.sort()
      ).toEqual(classes.sort());
    });

    it("gives an error when classes are provided as one string", () => {
      const classes = "classOne classTwo";
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<InputLabel classes={classes} />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("color prop", () => {
    it("applies color class to the label element with a valid color", () => {
      const color = "black";
      const child = <div>it's a label</div>;
      const hint = <span>it's a hint</span>;

      const wrapper = mount(
        <InputLabel color={color} children={child} hint={hint} />
      );

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .hasClass(color)
      ).toEqual(true);
    });

    it("applies color class to the InputHint element with a valid color", () => {
      const color = "black";
      const child = <div>it's a label</div>;
      const hint = <span>it's a hint</span>;

      const wrapper = mount(
        <InputLabel color={color} children={child} hint={hint} />
      );

      expect(
        wrapper
          .find("InputLabel")
          .find("InputHint")
          .props().color
      ).toEqual(color);
    });

    it("gives an error when given an invalid color", () => {
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<InputLabel color="red" />);

      expect(errors).toHaveBeenCalled();
    });
  });

  describe("hint prop", () => {
    it("displays given hint node inside the InputHint element", () => {
      const hint = <span>it's a hint</span>;

      const wrapper = mount(<InputLabel hint={hint} />);

      expect(
        wrapper
          .find("InputLabel")
          .find("InputHint")
          .contains(hint)
      ).toEqual(true);
    });

    it("does not render InputHint element when hint is not given", () => {
      const wrapper = mount(<InputLabel />);

      expect(wrapper.find("InputLabel").find("InputHint")).toHaveLength(0);
    });
  });

  describe("required prop", () => {
    it("displays an asterisk at the end of the label when required is true", () => {
      const child = <div>it's a label</div>;
      const required = true;

      const wrapper = mount(
        <InputLabel children={child} required={required} />
      );

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .text()
          .substr(-1)
      ).toEqual("*");
    });

    it("does not display an asterick when required is false", () => {
      const child = <div>it's a label</div>;
      const required = false;

      const wrapper = mount(
        <InputLabel children={child} required={required} />
      );

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .text()
          .substr(-1)
      ).not.toEqual("*");
    });
  });

  describe("weight prop", () => {
    it("applies weight class name on label element with a valid weight", () => {
      const weight = "bold";
      const child = <div>it's a label</div>;

      const wrapper = mount(<InputLabel children={child} weight={weight} />);

      expect(
        wrapper
          .find("InputLabel")
          .find("label")
          .hasClass(weight)
      ).toEqual(true);
    });

    it("gives an error when given an invalid weight", () => {
      const weight = "heavy";
      const child = <div>it's a label</div>;
      const errors = jest.spyOn(console, "error").mockImplementation();

      mount(<InputLabel children={child} weight={weight} />);

      expect(errors).toHaveBeenCalled();
    });
  });
});
