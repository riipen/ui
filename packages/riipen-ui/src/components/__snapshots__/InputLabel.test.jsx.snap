// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputLabel> renders correct snapshot 1`] = `
<withClasses(InputLabel)>
  <InputLabel
    classes={
      Array [
        "riipen",
        "riipen-inputlabel",
      ]
    }
    color="default"
    required={false}
    weight="regular"
  >
    <JSXStyle
      dynamic={
        Array [
          "#747474",
          "Roboto, Helvetica, Arial, sans-serif",
          400,
          "0.00938em",
          15,
          "#fff",
          "#000",
          300,
          500,
          700,
        ]
      }
      id="2351364854"
    />
  </InputLabel>
</withClasses(InputLabel)>
`;
